lib/modules/datasource/docker/index.ts

```diff
@@ -852,29 +951,76 @@ export class DockerDatasource extends Datasource {
     const newTag = newValue ?? 'latest';
     let digest: string | null = null;
     try {
-      let manifestResponse = await this.getManifestResponse(
-        registryHost,
-        dockerRepository,
-        newTag,
-        'head'
-      );
-      if (manifestResponse) {
-        if (hasKey('docker-content-digest', manifestResponse.headers)) {
+      let architecture: string | null | undefined = null;
+      if (currentDigest) {
+        architecture = await this.getImageArchitecture(
+          registryHost,
+          dockerRepository,
+          currentDigest
+        );
+      }
+
+      let manifestResponse: HttpResponse | null = null;
+      if (!architecture) {
+        manifestResponse = await this.getManifestResponse(
+          registryHost,
+          dockerRepository,
+          newTag,
+          'head'
+        );
+
+        if (
+          manifestResponse &&
+          hasKey('docker-content-digest', manifestResponse.headers)
+        ) {
           digest =
             (manifestResponse.headers['docker-content-digest'] as string) ||
             null;
-        } else {
-          logger.debug(
-            { registryHost },
-            'Missing docker content digest header, pulling full manifest'
-          );
-          manifestResponse = await this.getManifestResponse(
-            registryHost,
-            dockerRepository,
-            newTag
-          );
+        }
+      }
+
+      if (
+        architecture ||
+        (manifestResponse &&
+          !hasKey('docker-content-digest', manifestResponse.headers))
+      ) {
+        logger.debug(
+          { registryHost, dockerRepository },
+          'Architecture-specific digest or missing docker-content-digest header - pulling full manifest'
+        );
+        manifestResponse = await this.getManifestResponse(
+          registryHost,
+          dockerRepository,
+          newTag
+        );
+
+        if (architecture && manifestResponse) {
+          const manifestList = JSON.parse(manifestResponse.body) as
+            | ImageList
+            | Image
+            | OciImageList
+            | OciImage;
+          if (
+            manifestList.schemaVersion === 2 &&
+            (manifestList.mediaType === MediaType.manifestListV2 ||
+              manifestList.mediaType === MediaType.ociManifestIndexV1 ||
+              (!manifestList.mediaType && hasKey('manifests', manifestList)))
+          ) {
+            for (const manifest of manifestList.manifests) {
+              if (manifest.platform['architecture'] === architecture) {
+                digest = manifest.digest;
+                break;
```

Comment:
```suggestion
                return manifest.digest;
```

can we early return here? I think  `manifest.digest` shold never missing here?

this way we don't need the additional condition below.

Reply:
Correct, returning here would be possible but it would skip logging of the finally returned digest in:

https://github.com/renovatebot/renovate/blob/1e299f2a278535c4dadb84f2033dc51130ef247d/lib/modules/datasource/docker/index.ts#L1027-L1029